---
title: "Single-cell RNA-seq using available data"
author: "Michael C. Jeziorski"
date: "26 Jun 2019"
output: 
      BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{Vignette Title}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width = 8, fig.height = 3, echo = FALSE)
```
# Source of this tutorial {.smaller}
**R package: simpleSingleCell**  
**Chapter: Analyzing single-cell RNA-seq data containing UMI counts**

* Aaron T. L. Lun  
      + Cancer Research UK Cambridge Institute, Li Ka Shing Centre, Robinson Way, Cambridge CB2 0RE, United Kingdom (CRUK)  
* Davis J. McCarthy  
      + EMBL European Bioinformatics Institute, Wellcome Genome Campus, Hinxton, Cambridge CB10 1SD, United Kingdom (EMBL), St Vincent's Institute of Medical Research, 41 Victoria Parade, Fitzroy, Victoria 3065, Australia  
* John C. Marioni  
      + CRUK, EMBL, Wellcome Trust Sanger Institute, Wellcome Genome Campus, Hinxton, Cambridge CB10 1SA, United Kingdom

# Advantages of single-cell RNA-seq

- Preserves heterogeneity of cell types in a tissue
- Can provide insight into the number of different cell types in a tissue
- Can distinguish previously unknown cellular subtypes
- Offers opportunity for more targeted analysis of cells

# SingleCellExperiment

- S4 class available from Bioconductor for single-cell RNAseq data
- To install for Bioconductor version 3.8 or later (released October 2018; works with R version 3.5.0 or later):
```{r eval = FALSE, echo = TRUE}
# run only once
if (!requireNamespace("BiocManager"))
     install.packages("BiocManager")
BiocManager::install("SingleCellExperiment")
BiocManager::install("simpleSingleCell", version = "3.8")
BiocManager::install("scater")
```

```{r style, echo = FALSE, results = 'hide', message = FALSE}
library(BiocStyle)
library(knitr)
opts_chunk$set(error = FALSE, message = FALSE, warning = FALSE)
opts_chunk$set(fig.asp = 1)
library(BiocFileCache)
bfc <- BiocFileCache("raw_data", ask = FALSE)
base.url <- file.path("https://storage.googleapis.com",
    "linnarsson-lab-www-blobs/blobs/cortex")
mRNA.path <- "raw_data/expression_mRNA_17-Aug-2014.txt"
names(mRNA.path) <- "BFC1"
mito.path <- "raw_data/expression_mito_17-Aug-2014.txt"
names(mito.path) <- "BFC2"
spike.path <- "raw_data/expression_spikes_17-Aug-2014.txt"
names(spike.path) <- "BFC3"
readFormat <- function(infile) { 
    # First column is empty.
    metadata <- read.delim(infile, stringsAsFactors = FALSE, header = FALSE, nrow = 10)[,-1] 
    rownames(metadata) <- metadata[,1]
    metadata <- metadata[,-1]
    metadata <- as.data.frame(t(metadata))

    # First column after row names is some useless filler.
    counts <- read.delim(infile, stringsAsFactors = FALSE, 
        header = FALSE, row.names = 1, skip = 11)[,-1] 
    counts <- as.matrix(counts)
    return(list(metadata = metadata, counts = counts))
}
endo.data <- readFormat(mRNA.path)
spike.data <- readFormat(spike.path)
mito.data <- readFormat(mito.path)
m <- match(endo.data$metadata$cell_id, mito.data$metadata$cell_id)
mito.data$metadata <- mito.data$metadata[m,]
mito.data$counts <- mito.data$counts[,m]
stopifnot(identical(endo.data$metadata$cell_id, spike.data$metadata$cell_id)) # should be the same.
stopifnot(all(endo.data$metadata$cell_id ==  mito.data$metadata$cell_id)) # should now be the same.
raw.names <- sub("_loc[0-9]+$", "", rownames(endo.data$counts))
new.counts <- rowsum(endo.data$counts, group = raw.names, reorder = FALSE)
endo.data$counts <- new.counts
library(SingleCellExperiment)
all.counts <- rbind(endo.data$counts, mito.data$counts, spike.data$counts)
sce <- SingleCellExperiment(list(counts = all.counts), colData = endo.data$metadata)
nrows <- c(nrow(endo.data$counts), nrow(mito.data$counts), nrow(spike.data$counts))
is.spike <- rep(c(FALSE, FALSE, TRUE), nrows)
is.mito <- rep(c(FALSE, TRUE, FALSE), nrows)
isSpike(sce, "Spike") <- is.spike
library(org.Mm.eg.db)
ensembl <- mapIds(org.Mm.eg.db, keys = rownames(sce), keytype = "SYMBOL", column = "ENSEMBL")
rowData(sce)$ENSEMBL <- ensembl
library(scater)
sce <- calculateQCMetrics(sce, feature_controls = list(Mt = is.mito))
```

# Sample single-cell dataset

**Cell types in the mouse cortex and hippocampus revealed by single-cell RNA-seq**
  
Amit Zeisel, Ana B. Muñoz Manchado, Peter Lönnerberg, Gioele La Manno, Simone Codeluppi, Anna Juréus, Sueli Marques, Hermany Munguba, Liqun He, Christer Betsholtz, Charlotte Rolny, Gonçalo Castelo-Branco, Jens Hjerling-Leffler and Sten Linnarsson  
  
*Science* 347:1138-1142, 2015

http://linnarssonlab.org/cortex/

# Preparation of data

- Import data from Linnarsson Lab website
      + Reads are already filtered for quality and mapped
      + Data include mRNA, mitochondrial genes, and ERCC spike-ins (control)
- Clean up data and combine datasets
- Add ENSEMBL gene names
- Create `sce` object that contains count matrices and metadata

# Quality control overview
```{r libplotbrain}
par(mfrow = c(2,2), mar = c(5.1, 4.1, 0.1, 0.1))
hist(sce$total_counts/1e3, xlab = "Library sizes (thousands)", main = "", 
    breaks = 20, col = "grey80", ylab = "Number of cells")
hist(sce$total_features_by_counts, xlab = "Number of expressed genes", main = "", 
    breaks = 20, col = "grey80", ylab = "Number of cells")
hist(sce$pct_counts_Spike, xlab = "ERCC proportion (%)",
    ylab = "Number of cells", breaks = 20, main = "", col = "grey80")
hist(sce$pct_counts_Mt, xlab = "Mitochondrial proportion (%)", 
    ylab = "Number of cells", breaks = 20, main = "", col = "grey80")
```

```{r}
libsize.drop <- isOutlier(sce$total_counts, nmads = 3, type = "lower", log = TRUE)
feature.drop <- isOutlier(sce$total_features_by_counts, nmads = 3, type = "lower", log = TRUE)
spike.drop <- isOutlier(sce$pct_counts_Spike, nmads = 3, type = "higher")
sce <- sce[,!(libsize.drop | feature.drop | spike.drop)]
data.frame(ByLibSize = sum(libsize.drop), ByFeature = sum(feature.drop), 
    BySpike = sum(spike.drop), Remaining = ncol(sce))
```

# Gene-level metrics
```{r topgenebrain}
library(tidyverse)
fontsize <- theme(axis.text = element_text(size = 8), axis.title = element_text(size = 16))
geneplot <- plotHighestExprs(sce, n = 50) + fontsize
geneplot
```

## Gene abundance
```{r abhistbrain}
ave.counts <- calcAverage(sce, use_size_factors = FALSE)
hist(log10(ave.counts), breaks = 100, main = "", col = "grey",
    xlab = expression(Log[10] ~ "average count"))
rowData(sce)$ave.count <- ave.counts
to.keep <- ave.counts > 0
sce <- sce[to.keep,]
```

## Normalization of cell-specific biases
- An average count threshold of 0.1 is used to define high-abundance genes  
```{r normplotbrain}
library(scran)
set.seed(1000)
clusters <- quickCluster(sce, min.mean = 0.1, method = "igraph")
sce <- computeSumFactors(sce, cluster = clusters, min.mean = 0.1)
plot(sizeFactors(sce), sce$total_counts/1e3, log = "xy",
    ylab = "Library size (thousands)", xlab = "Size factor")
sce <- computeSpikeFactors(sce, type = "Spike", general.use = FALSE)
sce <- normalize(sce)
```

## Modeling and removing technical noise
```{r hvgplotbrain}
var.fit <- trendVar(sce, parametric = TRUE, loess.args = list(span = 0.4))
var.out <- decomposeVar(sce, var.fit)
plot(var.out$mean, var.out$total, pch = 16, cex = 0.6, xlab = "Mean log-expression", 
    ylab = "Variance of log-expression")
points(var.out$mean[isSpike(sce)], var.out$total[isSpike(sce)], col = "red", pch = 16)
curve(var.fit$trend(x), col = "dodgerblue", add = TRUE, lwd = 2)
```

## Distribution of expression of most highly expressed genes
```{r hvgvioplotbrain}
chosen.genes <- order(var.out$bio, decreasing = TRUE)[1:10]
plotExpression(sce, rownames(var.out)[chosen.genes], 
    point_alpha = 0.05, jitter_type = "jitter") + fontsize
set.seed(1000)
sce <- denoisePCA(sce, technical = var.fit$trend, approximate = TRUE)
```

## Dimensionality reduction
```{r tsneplotbrain, fig.wide}
set.seed(1000)
sce <- runTSNE(sce, use_dimred = "PCA", perplexity = 50)
tsne1 <- plotTSNE(sce, colour_by = "Neurod6") + fontsize
tsne2 <- plotTSNE(sce, colour_by = "Mog") + fontsize
multiplot(tsne1, tsne2, cols = 2)
```

# Clustering cells into subpopulations

## Graph-based clustering
```{r tsneclusterbrain, message = FALSE}
snn.gr <- buildSNNGraph(sce, use.dimred = "PCA")
cluster.out <- igraph::cluster_walktrap(snn.gr)
my.clusters <- cluster.out$membership
table(my.clusters)
sce$cluster <- factor(my.clusters)
plotTSNE(sce, colour_by = "cluster") + fontsize
```

## Evaluating graph-based clustering
```{r graphbrain}
mod.out <- clusterModularity(snn.gr, my.clusters, get.values = TRUE)
ratio <- mod.out$observed/mod.out$expected
lratio <- log10(ratio + 1)
cluster.gr <- igraph::graph_from_adjacency_matrix(ratio, 
    mode = "undirected", weighted = TRUE, diag = FALSE)
plot(cluster.gr, edge.width = igraph::E(cluster.gr)$weight*10)  
```

# Detecting subpopulation-specific marker genes
```{r, echo = FALSE, results = "hide"}
old.digits <- options()$digits
options(digits = 3)
```
```{r}
markers <- findMarkers(sce, my.clusters, direction = "up")
marker.set <- markers[["4"]]
head(marker.set[,1:8], 10) # only first 8 columns, for brevity
```
```{r, echo = FALSE, results = "hide"}
# Checking the cluster is what we wanted, along with cluster 10 ( = 9 in marker.set).
gad1 <- sapply(marker.set["Gad1",-(1:3)], sign)
stopifnot(gad1[9] == -1)
stopifnot(all(gad1[-9] == 1))

gad2 <- sapply(marker.set["Gad2",-(1:3)], sign)
stopifnot(gad2[9] == -1)
stopifnot(all(gad2[-9] == 1))

stopifnot(all(sapply(marker.set["Synpr",-(1:3)], sign) == 1))

options(digits = old.digits)
```

## Expression heatmap
```{r heatmapmarkerbrain, fig.wide = TRUE}
top.markers <- rownames(marker.set)[marker.set$Top <=  10]
plotHeatmap(sce, features = top.markers, columns = order(my.clusters),
    colour_columns_by = "cluster", cluster_cols = FALSE, 
    center = TRUE, symmetric = TRUE, zlim = c(-5, 5))
```